### ISSUE TITLE
option to collapse/expand namespaces

### ISSUE THREAD

**Issue by jonschlinkert** [node-base/base-namespace#1]
To make debug statements more readable, we should consider "collapsing" the namespaces by default, like:

``` sh
# expanded
base:templates:assemble:generate:verb:helpers

# collapsed
base...verb:helpers
```

then we can add support for an option to "expand" the namespace

**Comment by tunnckoCore** [comment]
Hmm.. yea, that is very good idea! I also noticed that in examples, that this too long namespaces are ugly.

Maybe there is a question: On how much namespaces to collapse? Two, three?

**Comment by jonschlinkert** [comment]
I think we should just do something simple, like this (pseudo-code-ish):

``` js
var segs = this._namespace.split(':');
var first = segs[0];
var last = segs[segs.length - 1];

var displayName = first + '...' + last;
```

For this to work, we would need to ensure that only "application" namespaces are on `this._namespace` or whatever the property is. (maybe we can just use `this.namespace`, is that used elsewhere?) (edit: meaning that "child" namespaces wouldn't be on the property)

Also, I'm thinking `namespace` should be a getter/setter. but you might have addressed this

**Comment by tunnckoCore** [comment]
> Also, I'm thinking namespace should be a getter/setter.

Yea, thought for it, but now i think that it shouldn't.

> we would need to ensure that only "application" namespaces 

Mm, maybe some registry which namespaces on per app basis? I know that now you may say it is too much, but it would be better, and no so hard.

> edit: meaning that "child" namespaces wouldn't be on the property

No. I'm :-1: for this one. I think it should always contain all namespaces. In combination with above idea for namespaces registry the final result would be great!

For example. When debug plugin is called in templates, it will write created namespaces to the registry with key `templates`, when it plugin is called in `assemble` it would be with assemble key. So then we can easily detect which namespaces are for which app and show/collapse only them.

edit: if we create it this way we have
- "fullpath" namespace
- "per app basis" namespaces
- when can easily manipulate them and expose options to configure more things

**Comment by jonschlinkert** [comment]
> I think it should always contain all namespaces.

I think you misunderstood what I meant:

**"application" namespace**

```
base:templates:assemble:verb
```

**child namespace**

```
helper
```

IMHO it's a no-brainer to keep a reference to the parent (application) namespace

**Comment by jonschlinkert** [comment]
> Mm, maybe some registry

hopefully my previous comment cleared this up. you're already creating these, we just need an easy way to "append" child namespaces to `base:templates:assemble:verb:foo...` inside an app.

> now i think that it shouldn't.

it's probably not necessary

**Comment by tunnckoCore** [comment]
> I think you misunderstood what I meant:

Oo, i got it. Hm... that is almost implemented with `this._debugNamespace`, and we can just create one more property, say `this._childNamespace` (edit: again, when rethink it, i see it as some kind of "registry") which just removes the application namespace from `this._debugNamespace`.

**Comment by tunnckoCore** [comment]
> we just need an easy way to "append" child namespaces

hm.. maybe i'm missing something, but i think it is done. Any call of `this.debug` "appends" more and more "child" namespaces, it is free to user to "add" new and new namespaces.

**Comment by jonschlinkert** [comment]
> say `this._childNamespace` which just removes the application namespace from this._debugNamespace.

yeah I think that's the right idea. maybe just call the child namespace property `_suffix`, then `_debugNamespace` can be a getter or function or whatever that joins `_namespace` with `_suffix`.

**Comment by tunnckoCore** [comment]
> child namespace property _suffix

Mm, ookey :+1:

**Comment by jonschlinkert** [comment]
or.. we can rename `_namespace` to `_debugPrefix`, then `_debugNamespace` would join `_debugPrefix` and `_debugSuffix`

**Comment by jonschlinkert** [comment]
`var debugNamespace = _debugPrefix + _debugSuffix`

**Comment by doowb** [comment]
for the collapse/expand part of this... i think we only need to change the `.namespace` property on the actual `debug` method created by the `debug` library (after it's created). This allows users to still do `DEBUG=base:templates:*` but only see `base...verb`.

I like the `_debugPrefix + _debugSuffix` idea since the namespaces mainly for debugging/logging.

**Comment by tunnckoCore** [comment]
@jonschlinkert Yes! Great idea!
- `debugNamespace` - full
- `debugSuffix` - "child" namespace
- `debugPrefix` - "application" namespace

> @doowb:  i think we only need to change the .namespace property

yea, i was playing with that

**Comment by tunnckoCore** [comment]
Okey, namespace thing is done in https://github.com/node-base/base-debug/pull/3.

I'm starting to think that it started to looks (may looks) a little bit confusing what `*-debug` is and what `*-namespace` is. But really, it's not.
- `base-debug` is magician behind composing sub-namespaces and adding prefix and suffix things
- `base-namespace` job is only to automagically create (based on `this._name`) stable "application" namespace (a.k.a `this._namespace`, which we are going to replace with `this._debugPrefix` now), that won't be modified. it also will have options for collapse/expand.

**Comment by tunnckoCore** [comment]
@jonschlinkert can you push a boilerplate on which i'll start adding the stuff?

**Comment by jonschlinkert** [comment]
> this._namespace, which we are going to replace with this._debugPrefix now

actually, sorry complicate this. since we were in a different repo when we spoke about this, I wasn't thinking about what this lib specifically should do, I was focused on the `base-debug`. But I think this lib should have no association with `debug`. it should just ensure that `this._namespace` is correct. 

then we can use this anywhere

**Comment by jonschlinkert** [comment]
> can you push a boilerplate on which i'll start adding the stuff?

sure, I'll try to do that today

**Comment by tunnckoCore** [comment]
> it should just ensure that this._namespace is correct.

Exactly what i'm talking about with that _"job is only to automagically create (based on this._name) stable "application" namespace"_, as it is currently.
Both are totally decoupled.

> But I think this lib should have no association with debug

it not have.

**Comment by jonschlinkert** [comment]
> a.k.a this._namespace, which we are going to replace with this._debugPrefix now

what did this mean then? maybe I'm misunderstanding

**Comment by tunnckoCore** [comment]
Haha, yea. I thought this sentence to be "which may be replaced with".
But nevermind, it's not a problem to save `this._namespace`. :+1:
